File: python/array/second_largest.py
def second_largest_naive(numbers: list[int]) -> int:
    # Deduplicate so equal copies of the largest don't shadow the true second
    # distinct maximum, then sort ascending. The second-largest is at index -2.
    # Time: O(n log n)  Space: O(n)
    unique_numbers: list[int] = sorted(set(numbers))
    return unique_numbers[-2] if len(unique_numbers) >= 2 else -1


def second_largest_two_pass(numbers: list[int]) -> int:
    # Time: O(n)  Space: O(1)
    numbers_length: int = len(numbers)
    if numbers_length < 2:
        return -1

    # Pass 1: find the largest value.
    first_largest: int = max(numbers)

    # Pass 2: find the largest value that is strictly less than first_largest.
    # -inf ensures any real integer immediately beats it on the first hit.
    second_largest: int | float = float("-inf")
    number: int
    for number in numbers:
        if number != first_largest and number > second_largest:
            second_largest = number

    # If second_largest was never updated, all elements are equal — no valid answer.
    return second_largest if second_largest != float("-inf") else -1


def second_largest_optimal(numbers: list[int]) -> int:
    # Single pass: maintain the top-two distinct values simultaneously.
    # Time: O(n)  Space: O(1)
    numbers_length: int = len(numbers)
    if numbers_length < 2:
        return -1

    first_largest: int | float = float("-inf")
    second_largest: int | float = float("-inf")
    number: int
    for number in numbers:
        if number > first_largest:
            # New overall maximum: demote the old first to second.
            second_largest = first_largest
            first_largest = number
        elif first_largest > number > second_largest:
            # Not a new maximum but beats the current second — update second only.
            second_largest = number

    return second_largest if second_largest != float("-inf") else -1


def main() -> None:
    numbers: list[int] = list(map(int, input().split(" ")))

    # Naive approach
    second_largest: int = second_largest_naive(numbers)
    print(second_largest)

    # Two-pass approach
    second_largest = second_largest_two_pass(numbers)
    print(second_largest)

    # Optimal approach
    second_largest = second_largest_optimal(numbers)
    print(second_largest)


if __name__ == "__main__":
    main()

File: rust/src/bin/array/second_largest.rs
/*
Test Cases
10 10 9
5 1 5 2 3 2
1 2

Output (naive / two-pass / optimal per input line)
9
9
9
*/

use std::io::{self, BufRead};

// Sort ascending, then scan backwards past all copies of the largest
// to find the first strictly smaller value — the second distinct maximum.
// Time: O(n log n)  Space: O(n)
fn second_largest_naive(numbers: &[i64]) -> i64 {
    if numbers.len() < 2 {
        return -1;
    }
    let mut sorted: Vec<i64> = numbers.to_vec();
    sorted.sort();
    let largest: i64 = *sorted.last().unwrap();
    for &val in sorted.iter().rev().skip(1) {
        if val != largest {
            return val;
        }
    }
    -1
}

// Two-pass: find max first, then scan for the largest value strictly below it
fn second_largest_two_pass(numbers: &[i64]) -> i64 {
    if numbers.len() < 2 {
        return -1;
    }
    let largest = *numbers.iter().max().unwrap();
    let mut second_largest = i64::MIN;
    for &number in numbers {
        if number != largest && number > second_largest {
            second_largest = number;
        }
    }
    if second_largest == i64::MIN {
        -1
    } else {
        second_largest
    }
}

// Optimal: single pass tracking first and second distinct maximums
fn second_largest_optimal(numbers: &[i64]) -> i64 {
    if numbers.len() < 2 {
        return -1;
    }
    let mut first_largest = i64::MIN;
    let mut second_largest = i64::MIN;
    for &number in numbers {
        if number > first_largest {
            second_largest = first_largest;
            first_largest = number;
        } else if first_largest > number && number > second_largest {
            second_largest = number;
        }
    }
    if second_largest == i64::MIN {
        -1
    } else {
        second_largest
    }
}

fn main() {
    let stdin = io::stdin();
    let line = stdin.lock().lines().next().unwrap().unwrap();
    let numbers: Vec<i64> = line
        .split_whitespace()
        .map(|s| s.parse().unwrap())
        .collect();

    println!("{}", second_largest_naive(&numbers));
    println!("{}", second_largest_two_pass(&numbers));
    println!("{}", second_largest_optimal(&numbers));
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_naive() {
        assert_eq!(second_largest_naive(&[10, 10, 9]), 9);
        assert_eq!(second_largest_naive(&[5, 1, 5, 2, 3, 2]), 3);
        assert_eq!(second_largest_naive(&[1, 2]), 1);
        assert_eq!(second_largest_naive(&[3, 1, 4, 1, 5, 9, 2, 6]), 6);
    }

    #[test]
    fn test_two_pass() {
        assert_eq!(second_largest_two_pass(&[10, 10, 9]), 9);
        assert_eq!(second_largest_two_pass(&[5, 1, 5, 2, 3, 2]), 3);
        assert_eq!(second_largest_two_pass(&[1, 2]), 1);
        assert_eq!(second_largest_two_pass(&[3, 1, 4, 1, 5, 9, 2, 6]), 6);
    }

    #[test]
    fn test_optimal() {
        assert_eq!(second_largest_optimal(&[10, 10, 9]), 9);
        assert_eq!(second_largest_optimal(&[5, 1, 5, 2, 3, 2]), 3);
        assert_eq!(second_largest_optimal(&[1, 2]), 1);
        assert_eq!(second_largest_optimal(&[3, 1, 4, 1, 5, 9, 2, 6]), 6);
    }
}
